pkg/dockerregistry/server/auth.go

```diff
@@ -272,6 +273,59 @@ func verifyOpenShiftUser(ctx context.Context, client client.UsersInterface) erro
 	return nil
 }
 
+// authorizeRepositoryAccess returns nil if the given client is granted access to the given
+// <namespace>/<imageRepo> repository. Otherwise the access is denied.
+func authorizeRepositoryAccess(ctx context.Context, namespace, imageRepo, verb string, client client.LocalSubjectAccessReviewsNamespacer) error {
+	if isCrossRepoMountFromSourceRepository(ctx, namespace, imageRepo, verb) {
+		return nil
+	}
+
+	return verifyImageStreamAccess(ctx, namespace, imageRepo, verb, client)
+}
+
+// isCrossRepoMountFromSourceRepository handles a case of accessing source repository requested by a cross
+// repository mount during a blob upload. During a push with the cross-repo mount, there are two authorization
+// checks:
+//
+//   1. repository:<source-namespace>/<source-name>:pull
+//   2. repository:<target-namespace>/<target-name>:push
+//
+// This function always grants access in the first case because OpenShift registry treats all the local blobs
+// as accessible once authorized to access given repository (target one).
+func isCrossRepoMountFromSourceRepository(ctx context.Context, namespace, imageRepo, verb string) bool {
```

Comment:
It's upstream's responsibility to do auth check for both repositories (source and target). See here: https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/github.com/docker/distribution/registry/handlers/app.go#L771. The source check is done first. The second check will decide whether the user is authorized to push.